feat: Add invariant testing (#760)
foundry-rs/book: src/forge/invariant-testing.md

# Invariant Testing
## Overview

Invariant testing allows for a set of mathematical invariants to be tested against randomized sequences of pre-defined function calls from pre-defined contracts. After each function call is performed, all defined invariants are asserted.

Invariant testing is a powerful tool to expose incorrect logic in protocols. Due to the fact that function call sequences are randomized and have fuzzed inputs, invariant testing can expose false assumptions and incorrect logic in edge cases and highly complex protocol states.

Invariant testing campaigns have two dimensions, `runs` and `depth`.
- `runs`: Number of times that a sequence of function calls is generated and run.

> - `depth`: Number of function calls made in a given `run`. All defined invariants are asserted after each function call is made.

Reviewer: Let's mention how reverted calls are treated within the `depth` counter (I think they are included and still increase depth when set to not fail on reverts?)

Author: Yeah can mention that

## Target Contracts

Target contracts are the set of contracts that will be called over the course of a given invariant test fuzzing campaign.

### Function Call Probability Distribution

Functions from these contracts will be called at random with fuzzed inputs. The probability of a function being called is broken down by contract and then by function.

For example:

```
targetContract1: 50%
├─ function1: 50% (25%)
└─ function2: 50% (25%)

targetContract2: 50%
├─ function1: 25% (12.5%)
├─ function2: 25% (12.5%)
├─ function3: 25% (12.5%)
└─ function4: 25% (12.5%)
```

> This is something to be mindful of when designing target contracts, as target contracts with less functions will have each function called more often due to this probability distribution.

Reviewer: We should really fix this, cc @gakonst https://github.com/foundry-rs/foundry/issues/2986

Reviewer 2: ACK

### Target Contract Setup

Target contracts can be set up using the following three methods:
1. Contracts that are deployed in the `setUp` function are automatically added to the set of target contracts.
2. Contracts that are added to the the `targetContracts` array are added to the set of target contracts. (TODO: check if this is possible during invariant campaigns yet)
3. Contracts that are deployed in the `setUp` can be **removed** from the target contracts if they are added to the `excludeContracts` array.

## Defining Invariants

Invariants are mathematical expressions that should always hold true over the course of a fuzzing campaign. There are different ways to assert invariants:

<table>
<tr><th>Type</th><th>Explanation</th><th>Example</th></tr>

<tr>

<td>Direct assertions</td>
<td>Directly query a protocol smart contracts and assert values are expected.</td>
<td>

```solidity
assertGe(
    token.totalAssets(),

>     token.totalSupply())

Reviewer: ```suggestion
    token.totalSupply()
```

)
```
</td>

</tr>

<tr>

<td>Ghost variable assertions</td>
<td>Query a protocol smart contract and compare it against a value that has been persisted in the test environment (ghost variable).</td>
<td>

```solidity
assertEq(
    token.totalSupply(),
    sumBalanceOf
)
```
</td>

</tr>

<tr>

<td>Naive implementation assertions</td>

> <td>Query the a protocol smart contract function and compare it against a naive and typically highly gas-inefficient implementation of the same desired logic.</td>

Reviewer 3: Typo: "the a".

<td>

```solidity
assertEq(
    pool.outstandingInterest(),
    test.naiveInterest()
)
```
</td>

</tr>
</table>

### Conditional Invariants

> Invariants must hold for the course over the course of a given fuzzing campaign, but that doesn't mean they must hold true in every situation. There is the possibility for certain invariants to be introduced/removed in a given scenario (e.g., during a liquidation). For this a dedicated testing contract should be used.

Reviewer 3: We should delete "for the course" here.

Author: Fixed thanks

> It is not recommended to introduce conditional logic into invariant assertions because they have the possibility of introducing false positives because of an incorrect code path.

Reviewer: Can you expand on this, maybe with an example?

## Target Contracts

Target contracts are the set of contracts that will be called over the course of a given invariant test fuzzing campaign. This set of contracts defaults to all contracts that were deployed in the `setUp` function, but can be customized to allow for more advanced invariant testing.

### Function Call Probability Distribution

Functions from these contracts will be called at random with fuzzed inputs. The probability of a function being called is broken down by contract and then by function.

For example:

```
targetContract1: 50%
├─ function1: 50% (25%)
└─ function2: 50% (25%)

targetContract2: 50%
├─ function1: 25% (12.5%)
├─ function2: 25% (12.5%)
├─ function3: 25% (12.5%)
└─ function4: 25% (12.5%)
```

This is something to be mindful of when designing target contracts, as target contracts with less functions will have each function called more often due to this probability distribution.

### Target Contract Setup

Target contracts can be set up using the following three methods:
1. Contracts that are deployed in the `setUp` function are automatically added to the set of target contracts.

> 2. Contracts that are added to the the `targetContracts` array are added to the set of target contracts. (TODO: check if this is possible during invariant campaigns yet)

Reviewer 4: The current wording is rather confusing (at least for me) - can you please formulate it in a way that people understand that if `targetContracts` method is present on a contract, forge will use the array returned by that method **instead** (and not all contracts in the `setUp` function).

Author: Is this the case? I haven't tested this. Updated the wording though let me know if that's better.

Reviewer 4: Well @msd1 told me that yesterday :) so maybe he should confirm as well.

Reviewer 3: @lucas-manuel and @pcaversaccio, can confirm that by using `targetContract`, the default behavior of Forge to automatically add all contracts deployed in the `setUp` function is turned off.

Reviewer 4: 👍 Awesome - I also tested it and confirm as well - at least with the latest nightly build version.

# Invariant Testing
## Overview

Invariant testing allows for a set of mathematical invariants to be tested against randomized sequences of pre-defined function calls from pre-defined contracts. After each function call is performed, all defined invariants are asserted.

Invariant testing is a powerful tool to expose incorrect logic in protocols. Due to the fact that function call sequences are randomized and have fuzzed inputs, invariant testing can expose false assumptions and incorrect logic in edge cases and highly complex protocol states.

Invariant testing campaigns have two dimensions, `runs` and `depth`.
- `runs`: Number of times that a sequence of function calls is generated and run.

> - `depth`: Number of function calls made in a given `run`. All defined invariants are asserted after each function call is made. If a function call reverts, the `depth` counter still increments.

Reviewer: @PaulRBerg do you think this explanation is clear? I know there was some confusion in https://github.com/foundry-rs/foundry/issues/3005 around what a `run` is

Reviewer 3: Ye, this is awesome.

Similar to how standard tests are run in Foundry by prefixing a function name with `test`, invariant tests are denoted by prefixing the function name with `invariant` (e.g., `function invariant_A()`).

## Defining Invariants

> Invariants are mathematical expressions that should always hold true over the course of a fuzzing campaign. A good invariant testing suite should have as many invariants as possible, and can have different testing suites for different protocol states.

Reviewer: They can also be things like "only the owner can do X" which isn't mathematical
 ```suggestion
Invariants are expressions that should always hold true over the course of a fuzzing campaign. A good invariant testing suite should have as many invariants as possible, and can have different testing suites for different protocol states.
```

Reviewer 3: I agree with @mds1 that we shouldn't define invariants as exclusively "mathematical expressions", but I don't think that example he gave is fitting in this context.

"only the owner can do X" sounds like a *function-level* invariant to me, whereas Foundry seems to me dealing with *system-level* invariants with its `invariant_` tests (the distinction is nicely explained by Anish Naik [here](https://youtu.be/QofNQxW_K08?t=2840)).

A better example would be time-series types of invariants, e.g. nonces that should only go up incrementally (as discussed in https://github.com/foundry-rs/foundry/issues/3005).

Author: Changed this to `defined assertions` to be more general, it gets defined more below.

> Invariants are mathematical expressions that should always hold true over the course of a fuzzing campaign. A good invariant testing suite should have as many invariants as possible, and can have different testing suites for different protocol states.

Reviewer 2: ```suggestion
Invariants are conditions expressions that should always hold true over the course of a fuzzing campaign. A good invariant testing suite should have as many invariants as possible, and can have different testing suites for different protocol states. Examples of an invariant would be "the xy=k formula always holds" for Uniswap, or "the sum of all owner balances is equal to the total supply" for an ERC20.
```

Author: Updated to this

There are different ways to assert invariants:

<table>
<tr><th>Type</th><th>Explanation</th><th>Example</th></tr>

<tr>

<td>Direct assertions</td>

> <td>Directly query a protocol smart contracts and assert values are expected.</td>

Reviewer: ```suggestion
<td>Directly query a protocol's smart contracts and assert values are expected.</td>
```

<td>

```solidity
assertGe(
    token.totalAssets(),
    token.totalSupply()
)
```
</td>

</tr>

<tr>

<td>Ghost variable assertions</td>
<td>Query a protocol smart contract and compare it against a value that has been persisted in the test environment (ghost variable).</td>
<td>

```solidity
assertEq(
    token.totalSupply(),
    sumBalanceOf
)
```
</td>

</tr>

<tr>

> <td>Naive implementation assertions</td>

Reviewer: Should we call this "Deoptimizing"? It matchings the RV talk's name and is shorter. Or we can write something like 
```suggestion
<td>Deoptimizing (Naive implementation assertions)</td>
```

Reviewer 2: i like

Author: @PaulRBerg https://www.youtube.com/watch?v=ETlNhV9jYJw&t=747s

Author: Update

## Target Contracts

Target contracts are the set of contracts that will be called over the course of a given invariant test fuzzing campaign. This set of contracts defaults to all contracts that were deployed in the `setUp` function, but can be customized to allow for more advanced invariant testing.

### Function Call Probability Distribution

Functions from these contracts will be called at random with fuzzed inputs. The probability of a function being called is broken down by contract and then by function.

For example:

```
targetContract1: 50%
├─ function1: 50% (25%)
└─ function2: 50% (25%)

targetContract2: 50%
├─ function1: 25% (12.5%)
├─ function2: 25% (12.5%)
├─ function3: 25% (12.5%)
└─ function4: 25% (12.5%)
```

This is something to be mindful of when designing target contracts, as target contracts with less functions will have each function called more often due to this probability distribution.

### Target Contract Setup

Target contracts can be set up using the following three methods:
1. Contracts that are deployed in the `setUp` function are automatically added to the set of target contracts.

> 2. Contracts that are added to the the `targetContracts` array are added to the set of target contracts and used instead of the contract deployed in the `setUp`.

Reviewer: Typo. Also are we certain this is true? Ref https://github.com/foundry-rs/foundry/issues/4187
```suggestion
2. Contracts that are added to the the `targetContracts` array are added to the set of target contracts and used instead of the contracts deployed in the `setUp`.
```

Reviewer 3: Ye, we are sure now that `targetContracts` overrides the default behavior to add all contracts from `setUp`.

https://github.com/foundry-rs/foundry/issues/4187#issuecomment-1406356194

> 3. Contracts that are deployed in the `setUp` can be **removed** from the target contracts if they are added to the `excludeContracts` array.

Reviewer: So really, the way you'd use these are:
- Option A: Pair (1) and (3)
- Option B: Use (2)
- Option C: More rare, all 3, if you need to manually add some already-existing contracts from e.g. a fork or `deployCode` usage

If that's correct, would be nice to give those sample use cases

Reviewer 3: I would bump up Option B and make it the first option mentioned - based on the experience I had with invariant testing so far, it seems to me that this will be the most common way invariant contracts will be targeted.

Reviewer 4: based on my personal previous experience with dapptools and now Foundry I can share the impression from @PaulRBerg.

Reviewer 2: +1 I would recommend B more so

Author: Moved around, lmk if this makes sense/should be reworded.

## Target Contract Patterns

This section will outline different approaches that can be used for invariant testing, as well as their pros and cons.

### Open Testing

The default configuration for target contracts is set to all contracts that are deployed during the setup. For smaller modules and more arithmetic contracts, this works well. For example:

```solidity
contract ExampleContract1 {

    uint256 val1;
    uint256 val2;
    uint256 val3;

    function addToA(uint256 amount) external {
        val1 += amount;
        val3 += amount;
    }

    function addToB(uint256 amount) external {
        val2 += amount;
        val3 += amount;
    }

}
```

This contract could be deployed and tested using the default target contract pattern:

```solidity
contract InvariantExample1 {

    ExampleContract1 foo;

    function setUp() external {
        foo = new ExampleContract1();
    }

    function invariant_A() external {
        assertEq(foo.val1() + foo.val2(), foo.val3());
    }

    function invariant_B() external {
        assertGe(foo.val1() + foo.val2(), foo.val1());
    }

}
```

> This setup will call `foo.addToA()` and `foo.addToB()` with a 50%-50% probability distribution with fuzzed inputs. Inevitably, the inputs will start to cause overflows and the function calls will start reverting. Since the default configuration in invariant testing is `fail_on_revert = true`, this will not cause the tests to fail. The invariants will hold throughout the rest of the fuzzing campaign and the result is that the test will pass. The output will look something like this:

Reviewer 5: Excellent documentation @lucas-manuel, thank you! I believe the default configuration here is `fail_on_revert = false` as per the [config README](https://github.com/foundry-rs/foundry/blob/master/config/README.md?plain=1#L183) which makes more sense in the context of this sentence and the final handler section.

Reviewer: Yea @lucas-manuel `false` is the default here

https://github.com/foundry-rs/foundry/blob/4718a390de2afc7372082d77afe7dc8d0e435215/config/src/invariant.rs#L31

```
[PASS] invariant_A() (runs: 50, calls: 10000, reverts: 5533)
[PASS] invariant_B() (runs: 50, calls: 10000, reverts: 5533)
```

### Invariant Handlers

For more complex and integrated protocols, more sophisticated target contract usage is required to achieve the desired results. For example, a ERC-4626 based contract that accepts deposits of another ERC-20 token:

```solidity
// SPDX-License-Identifier: UNLICENSED

> pragma solidity ^0.8.7;

Reviewer: May as well use a more recent version for all the pragmas
```suggestion
pragma solidity ^0.8.17;
```

Author: Updated

```solidity
    function deposit(uint256 assets) public virtual {
        asset.mint(address(this), assets);

        asset.approve(address(token), assets);

        uint256 shares = token.deposit(assets, address(this));
    }
```

This contract will provide the necessary setup before a function call is made in order to ensure it is successful.

Building on this concept, Handlers can be used to develop more sophisticated invariant tests. With Open invariant testing, the tests run as shown in the diagram below, with random sequences of function calls being made to the protocol contracts directly with fuzzed parameters. This will cause reverts for more complex systems as outlined above.

![Blank diagram](https://user-images.githubusercontent.com/44272939/214752968-5f0e7653-d52e-43e6-b453-cac935f5d97d.svg)

> By using Handler contracts and **excluding** all protocol contracts from the `targetContracts` array by using the `excludeContracts` function, all function calls made to protocol contracts can be made in a way that is governed by the Handler to ensure successful calls. This is outlined in the diagram below.

Reviewer: This sentence contradicts the sentence near https://github.com/foundry-rs/book/pull/760/files#r1088430050, i.e. if you use `targetContracts` do you need to also exclude the ones from setup?

Reviewer 3: Nope, you don't.

https://github.com/foundry-rs/foundry/issues/4187#issuecomment-1406356194

Author: Nice TIL! Updated locally

![Invariant Diagrams - Page 2](https://user-images.githubusercontent.com/44272939/214752977-053f60e6-c644-42a9-8cff-4fd85a2517ac.svg)

With this layer between the fuzzer and the protocol, more powerful testing can be achieved.

Within Handlers, "ghost variables" can be tracked across multiple function calls to add additional info for invariant tests. A good example of this is summing all of the `shares` that each LP owns after depositing into the ERC-4626 token as shown above, and using that in the invariant (`totalSupply == sumBalanceOf`).

> Another benefit is the ability to perform assertions on function calls as they are happening. An example is asserting the ERC-20 balance of the LP has decremented by `assets` during the `deposit` function call. In this way, handler functions are similar to fuzz tests because they can take in fuzzed inputs, perform state changes, and assert before/after state.

Reviewer: Demonstrating some patterns like `useRandomUser` and modifying block properties through handler methods would be great to include in this section

Reviewer 3: Example modifier:

```solidity
address[] public actors;
address internal currentActor;

modifier useActor(uint256 actorIndexSeed) {
    currentActor = actors[bound(actorIndexSeed, 0, actors.length - 1)];
    vm.startPrank(currentActor);
    _;
    vm.stopPrank();
}
```

Reviewer 2: let's add paul's modifier above^

> In addition, with Handlers, input parameters can be bounded to reasonable expected values such that `fail_on_revert` can be set to `true`. This ensures that every function call that is being made by the fuzzer must be successful against the protocol in order to get tests to pass. This is very useful for visibility and confidence that the protocol is being tested in the desired way.

Reviewer: Let's explicitly call out and link to the `bound` method for this input shaping. And hopefully soon `vm.assume` support too https://github.com/foundry-rs/foundry/issues/4190

# Invariant Testing
## Overview

Invariant testing allows for a set of invariant expressions to be tested against randomized sequences of pre-defined function calls from pre-defined contracts. After each function call is performed, all defined invariants are asserted.

Invariant testing is a powerful tool to expose incorrect logic in protocols. Due to the fact that function call sequences are randomized and have fuzzed inputs, invariant testing can expose false assumptions and incorrect logic in edge cases and highly complex protocol states.

> Invariant testing campaigns have two dimensions, `runs` and `depth`.

Reviewer 4: ```suggestion
Invariant testing campaigns have two dimensions, `runs` and `depth`:
```

### Function Call Probability Distribution

Functions from these contracts will be called at random with fuzzed inputs. The probability of a function being called is broken down by contract and then by function.

For example:

```
targetContract1: 50%
├─ function1: 50% (25%)
└─ function2: 50% (25%)

targetContract2: 50%
├─ function1: 25% (12.5%)
├─ function2: 25% (12.5%)
├─ function3: 25% (12.5%)
└─ function4: 25% (12.5%)
```

This is something to be mindful of when designing target contracts, as target contracts with less functions will have each function called more often due to this probability distribution.

### Invariant Test Helper Functions
Invariant test helper functions are included in [`forge-std`](https://github.com/foundry-rs/forge-std/blob/master/src/StdInvariant.sol) to allow for configurable invariant test setup. The helper functions are outlined below:

| Function | Description |
|-|-|
| `excludeContract(address newExcludedContract_)` | Adds a given address to the `_excludedContracts` array. This set of contracts is explicitly excluded from the target contracts.|

> | `excludeSender(address newExcludedSender_)` | Adds a given address to the to the `_excludedSenders` array. This set of addresses is explictly excluded from the target senders. |

Reviewer 4: ```suggestion
| `excludeSender(address newExcludedSender_)` | Adds a given address to the to the `_excludedSenders` array. This set of addresses is explicitly excluded from the target senders. |
```

> | `excludeArtifact(string memory newExcludedArtifact_)` | Adds a given string to the to the `_excludedArtifacts` array. This set of strings is explictly excluded from the target artifacts. |

Reviewer 4: ```suggestion
| `excludeArtifact(string memory newExcludedArtifact_)` | Adds a given string to the to the `_excludedArtifacts` array. This set of strings is explicitly excluded from the target artifacts. |
```

> | `targetArtifact(string memory newTargetedArtifact_)` | Adds a given string to the to the `_targetedArtifacts` array. This set of strings is used for the target artifacts.  |

Reviewer 3: Double `to the to the` typo.

| `targetArtifactSelector(FuzzSelector memory newTargetedArtifactSelector_)` | Adds a given `FuzzSelector` to the to `_targetedArtifactSelectors` array. This set of `FuzzSelector`s is used for the target artifact selectors. |
| `targetContract(address newTargetedContract_)` | Adds a given address to the to `_targetedContracts` array. This set of addresses is used for the target contracts. This array overwrites the set of contracts that was deployed during the `setUp`. |
| `targetSelector(FuzzSelector memory newTargetedSelector_)` | Adds a given `FuzzSelector` to the to `_targetedSelectors` array. This set of `FuzzSelector`s is used for the target contract selectors. |
| `targetSender(address newTargetedSender_)` | Adds a given address to the to `_targetedSenders` array. This set of addresses is used for the target senders. |

### Target Contract Setup

Target contracts can be set up using the following three methods:

> 1. Contracts that are manually added to the the `targetContracts` array are added to the set of target contracts.

Reviewer 3: Double `the the` typo.

## Defining Invariants

Invariants are conditions expressions that should always hold true over the course of a fuzzing campaign. A good invariant testing suite should have as many invariants as possible, and can have different testing suites for different protocol states.

Examples of invariants are:
- *"The xy=k formula always holds"* for Uniswap

> - *"The sum of all owner balances is equal to the total supply"* for an ERC-20 token.

Reviewer: Nit, since "owner" sounds like a privileged actor 
```suggestion
- *"The sum of all user balances is equal to the total supply"* for an ERC-20 token.
```
